### PR TITLE
support reservation

### PR DESCRIPTION
**Pull request by lzlaa**

### CODE REVIEW

File: cmd/scheduler/app/options/options.go [hunk 1]
```diff
@@ -168,6 +168,7 @@ func (o *Options) Flags() (nfs cliflag.NamedFlagSets) {
 	fs.Float32Var(&o.ComponentConfig.ClientConnection.QPS, "kube-api-qps", o.ComponentConfig.ClientConnection.QPS, "QPS to use while talking with kubernetes apiserver. This parameter overrides the value defined in config file, which is specified in --config.")
 	fs.Int32Var(&o.ComponentConfig.ClientConnection.Burst, "kube-api-burst", o.ComponentConfig.ClientConnection.Burst, "burst to use while talking with kubernetes apiserver. This parameter overrides the value defined in config file, which is specified in --config.")
 
+	fs.Int64Var(&o.ComponentConfig.ReservationTimeOutSeconds, "reservation-ttl", o.ComponentConfig.ReservationTimeOutSeconds, "how long resources will be reserved (for resource reservation).")
```
Comment (binacs) [inline cmd/scheduler/app/options/options.go:171]: nit: move this to #L185-L192 for more fine-grained configuration in the future

File: manifests/quickstart-feature-examples/godel-demo-default.yaml [hunk 2]
```diff
@@ -4,5 +4,9 @@ name: godel-demo-default
 nodes:
   - role: control-plane
     image: kindest/node:v1.21.1
+  - role: worker
```
Comment (binacs) [inline manifests/quickstart-feature-examples/godel-demo-default.yaml:7]: Do we need more workers for demo cluster?

File: pkg/binder/utils/reservation.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+		prr, err := prrLister.Reservations(ns).Get(name)
+		if err != nil {
+			klog.ErrorS(err, "get pod reservation request for updating, but failed", "prr", podutil.GetReservationKey(prr))
+			return err
+		}
+		newPrr := prr.DeepCopy()
+		// skip pod reservation request is not in reserved status.
+		if shouldSkipUpdateReservationStatus(prr) {
+			klog.V(4).InfoS("skip updating pod reservation request, pod reservation request is not in reserved status", "prr", podutil.GetReservationKey(prr))
```
Comment (binacs) [inline pkg/binder/utils/reservation.go:77]: nit: skip -> Skip

File: pkg/binder/utils/reservation.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+		if succeed {
+			newPrr.Status.CurrentOwners = schedulingv1a1.CurrentOwners{
+				Name:      assumedPod.Name,
+				Namespace: assumedPod.Namespace,
+				UID:       assumedPod.UID,
+			}
+			newPrr.Status.Phase = schedulingv1a1.ReservationMatched
+		} else {
+			// TODO: add to count and set in unavaliable status.
```
Comment (binacs) [inline pkg/binder/utils/reservation.go:90]: nit: no need to go on in this situation

File: pkg/scheduler/eventhandlers.go [hunk 5]
```diff
@@ -35,12 +30,17 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2"
 
+	nodev1alpha1 "github.com/kubewharf/godel-scheduler-api/pkg/apis/node/v1alpha1"
```
Comment (binacs) [inline pkg/scheduler/eventhandlers.go:33]: Should packages from other repos (godel-scheduler-api, katalyst-api) be located in above?

File: pkg/scheduler/framework/handle/handle.go [hunk 6]
```diff
@@ -84,4 +85,8 @@ type UnitFrameworkHandle interface {
 	FindStore(commonstore.StoreName) commonstore.Store
 
 	GetMaxWaitingDeletionDuration() time.Duration
+
+	// TODO: cleanup the following function methods.
```
Comment (binacs) [inline pkg/scheduler/framework/handle/handle.go:89]: Move theis line to #L87 considering that the GetMaxWaitingDeletionDuration should also be cleand up

File: pkg/util/resource_constants.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import v1 "k8s.io/api/core/v1"
+
+const (
+	ResourceBytedanceSocket v1.ResourceName = "bytedance.com/socket"
```
Comment (binacs) [inline pkg/util/resource_constants.go:22]: Please double check that whether bytedance.com may be claimed in open source projects

**Review by binacs — COMMENTED** [review]
Great work! LGTM overall just a few comments

File: manifests/quickstart-feature-examples/godel-demo-default.yaml (hunk 2 above)
Comment (lzlaa) [reply manifests/quickstart-feature-examples/godel-demo-default.yaml:7]: Yes, reservation needs two worker nodes at least, and one single worker node may cause unexpected error.

File: pkg/scheduler/framework/handle/handle.go (hunk 6 above)
Comment (lzlaa) [reply pkg/scheduler/framework/handle/handle.go:87]: fixed

File: pkg/binder/utils/reservation.go (hunk 4 above)
Comment (lzlaa) [reply pkg/binder/utils/reservation.go]: fixed

File: pkg/binder/utils/reservation.go (hunk 3 above)
Comment (lzlaa) [reply pkg/binder/utils/reservation.go]: fixed

File: cmd/scheduler/app/options/options.go (hunk 1 above)
Comment (lzlaa) [reply cmd/scheduler/app/options/options.go]: fixed

**Review by binacs — APPROVED** [review]
lgtm

File: docs/features/resource-reservation.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+The optional startup parameter `--reservation-ttl` is used to control the Reservation CR timeout (default value: '60', unit: seconds). If the Reservation CR is not matched after the timeout, the controller manager will delete the Reservation object to notify the scheduler to release the reserved resources.
+
+In addition, the global parameter `--reservation-ttl` can be set manually on the Pod or Deployment object annotation to override the controller manager's global parameter.
+
+```yaml
+godel.bytedance.com/reservation-ttl=7200
+```
+
+For example, if you configure the above annotation on a Pod or Deployment, the scheduler will automatically release the reserved resources after 24 hours if there is no consumer for the reserved resources.
```
Comment (binacs) [inline docs/features/resource-reservation.md:30]: What if both of the Deployment and Pod configured this field?

Priority needs to be clearly described.

File: docs/features/resource-reservation.md [hunk 9]
```diff
@@ -0,0 +1,231 @@
+# Resource Reservation User Documentation
+
+This document uses a kind cluster as an example to introduce how to enable and use the resource reservation capability of the Godel Scheduler.
+
+## Prerequisites
+
+Please make sure that the Godel Controller Manager component has been deployed in the kind cluster along with other scheduler components as control plane components.
+
+```shell
+$ k get po -n godel-system
```
[the remark this answers is not in the document]
Comment (lzlaa) [reply docs/features/resource-reservation.md]: ok, replace "k" to "kubectl".

File: docs/features/resource-reservation.md (hunk 8 above)
Comment (lzlaa) [reply docs/features/resource-reservation.md:30]: added some notes about the priority

**Review by binacs — DISMISSED** [review]
lgtm

File: cmd/controller/app/options/insecure_serving.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+	BindAddress string // overrides the structs above on ApplyTo, ignored on ApplyToFromLoadedConfig
+}
+
+// AddFlags adds flags for the insecure serving options.
+func (o *CombinedInsecureServingOptions) AddFlags(fs *pflag.FlagSet) {
+	if o == nil {
+		return
+	}
+
+	fs.StringVar(&o.BindAddress, "address", o.BindAddress, "DEPRECATED: the IP address on which to listen for the --port port (set to 0.0.0.0 for all IPv4 interfaces and :: for all IPv6 interfaces). See --bind-address instead.")
```
Comment (zryfish) [inline cmd/controller/app/options/insecure_serving.go:47]: Better to remove those deprecated flags.

File: cmd/controller/app/options/insecure_serving.go [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+			*addr = ""
+		} else {
+			*addr = net.JoinHostPort(is.BindAddress.String(), strconv.Itoa(is.BindPort))
+		}
+	}
+
+	return nil
+}
+
+func updateDeprecatedInsecureServingOptionsFromAddress(is *apiserveroptions.DeprecatedInsecureServingOptionsWithLoopback, addr string) error {
```
Comment (zryfish) [inline cmd/controller/app/options/insecure_serving.go:122]: This function is unnecessary if  deprecated flags are removed.

File: cmd/controller/app/options/leadermigration.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+	Enabled bool
+
+	// ControllerMigrationConfig is the path to the file of LeaderMigrationConfiguration type.
+	// It can be set with --leader-migration-config flag
+	// If the path is "" (default vaule), the default vaule will be used.
+	ControllerMigrationConfig string
+}
+
+// DefaultLeaderMigrationOptions returns a LeaderMigrationOptions with default values.
+func DefaultLeaderMigrationOptions() *LeaderMigrationOptions {
```
Comment (zryfish) [inline cmd/controller/app/options/leadermigration.go:41]: I don't think we need LeaderMigration. According to design doc https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2436-controller-manager-leader-migration, 

> Support a migration process for large scale and highly available Kubernetes clusters using the in-tree cloud providers (via kube-controller-manager and kubelet) to their out-of-tree equivalents (via cloud-controller-manager).

File: cmd/controller/app/options/reservation.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+	}
+	cfg.ReservationTTL = opt.ReservationTTL
+	cfg.ReservationCheckPeriod = opt.ReservationCheckPeriod
+	cfg.MatchedRequestExtraTTL = opt.MatchedRequestExtraTTL
+	cfg.IgnoredNamespace = opt.IgnoredNamespace
+	return nil
+}
+
+// TODO
+func (opt *ReservationControllerOptions) Validate() error {
```
Comment (zryfish) [inline cmd/controller/app/options/reservation.go:50]: It's better to add some validations.

File: docs/features/resource-reservation.md [hunk 14]
```diff
@@ -0,0 +1,234 @@
+# Resource Reservation User Documentation
+
+This document uses a kind cluster as an example to introduce how to enable and use the resource reservation capability of the Godel Scheduler.
+
+## Prerequisites
+
```
Comment (zryfish) [inline docs/features/resource-reservation.md:6]: But how to deploy godel controller manager, is there any documentation, or links maybe?

File: pkg/binder/controller/movement_controller.go [hunk 15]
```diff
@@ -1,3 +1,19 @@
+/*
+Copyright 2023 The Godel Scheduler Authors.
```
Comment (zryfish) [inline pkg/binder/controller/movement_controller.go:2]: LICENSE

File: cmd/controller/app/options/insecure_serving.go (hunk 10 above)
Comment (lzlaa) [reply cmd/controller/app/options/insecure_serving.go:47]: I think it is best to keep the behavior consistent with other scheduler components, retain the insecure server configuration, and clean it up later.

File: cmd/controller/app/options/insecure_serving.go (hunk 11 above)
Comment (lzlaa) [reply cmd/controller/app/options/insecure_serving.go:122]: > I think it is best to keep the behavior consistent with other scheduler components, retain the insecure server configuration, and clean it up later.

As the opinion shows, we need keep these flags.

File: docs/features/resource-reservation.md (hunk 14 above)
Comment (lzlaa) [reply docs/features/resource-reservation.md:6]: Follow the `docs/features/kind-cluster-setup.md`, users can deploy godel controller manager in kind cluster.

File: pkg/binder/controller/movement_controller.go (hunk 15 above)
Comment (lzlaa) [reply pkg/binder/controller/movement_controller.go]: Good catch. fix it.

File: cmd/controller/app/options/leadermigration.go (hunk 12 above)
Comment (lzlaa) [reply cmd/controller/app/options/leadermigration.go:41]: Good idea, removed it.

**Review by zryfish — APPROVED** [review]

**Pull request merged**